cpp/src/copying/concatenate.cu

```diff
@@ -329,58 +330,134 @@ std::unique_ptr<column> concatenate_dispatch::operator()<cudf::struct_view>()
 
 namespace {
 
+void bounds_and_type_check(host_span<column_view const> cols, rmm::cuda_stream_view stream);
+
+/**
+ * @brief Functor for traversing child columns and recursively verifying concatenation
+ * bounds and types.
+ */
+class traverse_children {
+ public:
+  // nothing to do for simple types.
+  template <typename T>
+  void operator()(host_span<column_view const>, rmm::cuda_stream_view)
+  {
+  }
+
+ private:
+  // verify length of concatenated offsets.
+  void check_offsets_size(host_span<column_view const> cols)
+  {
+    // offsets.  we can't just add up the total sizes of all offset child columns because each one
+    // has an extra value, regardless of the # of parent rows.  So we have to add up the total # of
+    // rows in the base column and add 1 at the end
+    size_t const total_offset_count =
+      std::accumulate(cols.begin(),
+                      cols.end(),
+                      std::size_t{},
+                      [](size_t a, auto const& b) -> size_t { return a + b.size(); }) +
+      1;
+    // note:  output text must include "exceeds size_type range" for python error handling
+    CUDF_EXPECTS(
+      total_offset_count <= static_cast<std::size_t>(std::numeric_limits<size_type>::max()),
+      "Total number of concatenated offsets exceeds size_type range");
+  }
+};
+
+template <>
+void traverse_children::operator()<cudf::string_view>(host_span<column_view const> cols,
+                                                      rmm::cuda_stream_view stream)
+{
+  // verify offsets
+  check_offsets_size(cols);
+
+  // chars
+  size_t const total_char_count = std::accumulate(
+    cols.begin(), cols.end(), std::size_t{}, [stream](size_t a, auto const& b) -> size_t {
+      strings_column_view scv(b);
+      return a + (b.is_empty()
+                    ? 0
+                    : cudf::detail::get_value<offset_type>(
+                        scv.offsets(), scv.offset() + b.size(), stream) -
+                        cudf::detail::get_value<offset_type>(scv.offsets(), scv.offset(), stream));
```

Comment:
Each `get_value` implicitly calls stream sync. So we have 2 syncs here. I wonder if we need to manually call `cudaMemcpy` twice, then justs do 1 stream sync for better performance.